vault/auth.go

```diff
@@ -820,6 +824,23 @@ func (c *Core) setupCredentials(ctx context.Context) error {
 			}
 		}
 
+		// Do not start up deprecated builtin plugins. If this is a major/minor
+		// upgrade, stop unsealing and shutdown. If we've already mounted this
+		// plugin, skip backend initialization and mount the data for posterity.
+		if versions.IsBuiltinVersion(entry.RunningVersion) {
+			shutdown := isMajorOrMinorUpgrade(c.currentVaultVersion.Version, entry.LastMounted)
+			_, err := c.handleDeprecatedMountEntry(ctx, entry, consts.PluginTypeCredential)
+			if shutdown && err != nil {
+				go c.ShutdownCoreError(err)
```

Comment:
> Vault simply loops over and over attempting to unseal.

It doesn't just seal itself?

Reply:
In my experience, no it does not. There may be something I'm doing wrong, but from my POV it has never behaved this way (since I started working on the deprecation unseal shutdown).

Reply:
Regarding the goroutine shutdown v. blocking shutdown, see below.

### goroutine shutdown:

<details>
<summary>logs</summary>

```
2022-12-08T14:34:01.952-0500 [INFO]  vault.core: pre-seal teardown starting
2022-12-08T14:34:01.952-0500 [ERROR] vault.core: shutting down core: error="could not mount \"app-id\": mount entry associated with pending removal builtin"
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: shutdown called
2022-12-08T14:34:01.952-0500 [INFO]  vault.core: stopping raft active node
2022-12-08T14:34:01.952-0500 [INFO]  vault.core: marked as sealed
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: clearing forwarding clients
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: done clearing forwarding clients
2022-12-08T14:34:01.952-0500 [INFO]  vault.core: pre-seal teardown complete
2022-12-08T14:34:01.952-0500 [ERROR] vault.core: post-unseal setup failed: error="failed to setup auth table"
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: finished triggering standbyStopCh for runStandby
2022-12-08T14:34:01.952-0500 [DEBUG] vault.storage.raft: starting leadership transfer: id=<nil> address=<nil>
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: shutting down periodic key rotation checker
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: shutting down periodic leader refresh
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: shutting down periodic metrics
2022-12-08T14:34:01.952-0500 [DEBUG] vault.storage.raft: cannot find peer
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: shutting down leader elections
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: stop channel triggered in runStandby
2022-12-08T14:34:01.952-0500 [DEBUG] vault.core: runStandby done
2022-12-08T14:34:01.952-0500 [ERROR] vault.storage.raft.raft-net: failed to accept connection: error="Raft RPC layer closed"
2022-12-08T14:34:01.952-0500 [INFO]  vault.core: stopping cluster listeners
2022-12-08T14:34:01.952-0500 [INFO]  vault.core.cluster-listener: forwarding rpc listeners stopped
2022-12-08T14:34:02.247-0500 [INFO]  vault.core.cluster-listener: rpc listeners successfully shut down
2022-12-08T14:34:02.247-0500 [INFO]  vault.core: cluster listeners successfully shut down
2022-12-08T14:34:02.247-0500 [DEBUG] vault.core: sealing barrier
2022-12-08T14:34:02.248-0500 [INFO]  vault.core: vault is sealed
```
</details>

<details>
<summary>client behavior</summary>

```bash
>  vault status
Key                Value
---                -----
Seal Type          shamir
Initialized        true
Sealed             true
Total Shares       5
Threshold          3
Unseal Progress    0/3
Unseal Nonce       n/a
Version            1.13.0-dev1
Build Date         n/a
Storage Type       raft
HA Enabled         true
```
</details>

### blocking shutdown:

<details>
<summary>vault logs</summary>

```
2022-12-08T14:40:34.644-0500 [INFO]  vault.core: successfully enabled credential backend: type=token version="" path=token/ namespace="ID: root. Path: "
2022-12-08T14:40:34.644-0500 [ERROR] vault.core: shutting down core: error="could not mount \"app-id\": mount entry associated with pending removal builtin"
2022-12-08T14:40:34.644-0500 [DEBUG] vault.core: shutdown called
2022-12-08T14:40:34.644-0500 [INFO]  vault.core: marked as sealed
2022-12-08T14:40:34.644-0500 [DEBUG] vault.core: clearing forwarding clients
2022-12-08T14:40:34.644-0500 [DEBUG] vault.core: done clearing forwarding clients
```
</details>

<details>
<summary>vault client behavior</summary>

```bash
>  vault status


^C⏎
```
</details>

Reply:
I was confused, the behaviour I had in mind is what we do in non-HA setups.  The looping we're doing is not in trying to unseal, but in trying to become active.

Reply:
Nick and I discussed this out of band and are convinced it is the correct behavior, so going to mark this resolved for now.